perun-engine/src/main/java/cz/metacentrum/perun/engine/scheduling/impl/BlockingGenExecutorCompletionService.java

```diff
@@ -25,65 +23,107 @@
  * @see cz.metacentrum.perun.engine.runners.GenCollector
  *
  * @author David Šarman
+ * @author Pavel Zlámal <zlamal@cesnet.cz>
  */
 public class BlockingGenExecutorCompletionService implements BlockingCompletionService<Task> {
 
 	private final static Logger log = LoggerFactory.getLogger(BlockingGenExecutorCompletionService.class);
 	private CompletionService<Task> completionService;
-	@Autowired
-	@Qualifier("generatingTasks")
-	private BlockingBoundedMap<Integer, Task> executingTasks;
-	private int limit;
+	private ConcurrentMap<Future<Task>, Task> executingGenTasks = new ConcurrentHashMap<>();
+	private Semaphore semaphore;
```

Comment:
Could you add a comment here, describing how the semaphore should work? It will be a huge help to anyone who will try to understand your code.